go/vt/schemamanager/schemaswap/schema_swap.go

```diff
@@ -75,6 +90,290 @@ type shardSchemaSwap struct {
+	for _, shard := range schemaSwap.allShards {
+		go shard.readShardMetadata(metadataChannel)
+	}
+	var initError error
+	var cntDataReceived int
+	for cntDataReceived != len(schemaSwap.allShards) {
+		metadata := <-metadataChannel
+		cntDataReceived++
+		if metadata.lastStartedSwap == metadata.lastFinishedSwap {
+			// The shard doesn't have schema swap started yet.
+			nextSwapID := metadata.lastFinishedSwap + 1
+			if schemaSwap.swapID == 0 {
+				schemaSwap.swapID = nextSwapID
+			} else if schemaSwap.swapID != nextSwapID {
+				initError = fmt.Errorf(
+					"Next schema swap id on shard %v should be %v, while for other shard(s) it should be %v",
+					metadata.shardName, nextSwapID, schemaSwap.swapID)
+			}
+		} else if metadata.lastStartedSwap < metadata.lastFinishedSwap {
+			initError = fmt.Errorf(
+				"Bad swap metadata on shard %v: LastFinishedSchemaSwap=%v is greater than LastStartedSchemaSwap=%v",
+				metadata.shardName, metadata.lastFinishedSwap, metadata.lastStartedSwap)
+		} else if schemaSwap.swapID != 0 && schemaSwap.swapID != metadata.lastStartedSwap {
+			initError = fmt.Errorf(
+				"Shard %v has an already started schema swap with an id %v, while for other shard(s) id should be equal to %v",
+				metadata.shardName, metadata.lastStartedSwap, schemaSwap.swapID)
+		} else if metadata.currentSQL != sql {
+			initError = fmt.Errorf("Shard %v has an already started schema swap with a different set of SQL statements", metadata.shardName)
+		} else {
+			schemaSwap.swapID = metadata.lastStartedSwap
+		}
+	}
+	close(metadataChannel)
+	if initError != nil {
+		return initError
+	}
+	initError = schemaSwap.runOnAllShards(
+		func(shard *shardSchemaSwap) error {
+			return shard.writeStartedSwap(sql)
+		})
+	return initError
+}
+
+// finalizeSwap finishes the completed swap process by verifying that it's actually complete
+// and then modifying the database to register the completion.
+func (schemaSwap *Swap) finalizeSwap() error {
+	err := schemaSwap.runOnAllShards(
+		func(shard *shardSchemaSwap) error {
+			return shard.verifySwapApplied()
+		})
+	if err != nil {
+		return err
+	}
+	err = schemaSwap.runOnAllShards(
+		func(shard *shardSchemaSwap) error {
+			return shard.writeFinishedSwap()
+		})
+	return err
+}
+
+// getMasterTablet returns the tablet that is currently master on the shard.
+func (shardSwap *shardSchemaSwap) getMasterTablet() (*topodatapb.Tablet, error) {
+	topoServer := shardSwap.parent.topoServer
+	shardInfo, err := topoServer.GetShard(shardSwap.parent.ctx, shardSwap.parent.keyspace, shardSwap.shard)
+	if err != nil {
+		return nil, err
+	}
+	tabletInfo, err := topoServer.GetTablet(shardSwap.parent.ctx, shardInfo.MasterAlias)
+	if err != nil {
+		return nil, err
+	}
+	return tabletInfo.Tablet, nil
+}
+
+// readShardMetadata reads info about schema swaps on this shard from _vt.shard_metadata table.
+func (shardSwap *shardSchemaSwap) readShardMetadata(metadataChannel chan *shardSwapMetadata) {
```

Comment:
I believe returning error here will hurt code compactness and readability. So I left it as is.